### PR TITLE
Sorry for the big PR

### PR DESCRIPTION
**Pull request by chuckbutler**
Cleanup layers with FlagManager
    - Cleanup syncthing botched transfer
    - Refactor service/env templates to follow upstream templates
    - Adds FlagManager to the master layer
    - Adds the application version declarations and fixes some bad imports
    - This normalizes the states so the charm behaves accordingly when no
      resources have been provided, and refactors the end-user messaging.
    - Adds the worker layer
    - mbruzek code-review comments cleanup
    - remove docker dependency on startup

### CODE REVIEW

File: cluster/juju/layers/kube-dns/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+```
+juju deploy kube-dns
+juju add-relation kube-dns mycharm
+```
+
+## More Information
+
+[Upstream Documentation](https://kubernetes.io/docs/admin/dns)
+
+
```
Comment (mbruzek) [inline cluster/juju/layers/kube-dns/README.md]: You will need more of a readme than this. It is ok for now, but the readme is really short.

File: cluster/juju/layers/kube-dns/config.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+options:
+  domain:
+    type: string
+    default: "cluster.local"
```
Comment (mbruzek) [inline cluster/juju/layers/kube-dns/config.yaml]: Does this need to be a list? In some of the kubernetes configuration scripts I see them setting the sans to different DNS names:
```
DNS:kubernetes,DNS:kubernetes.default,DNS:kubernetes.default.svc,DNS:kubernetes.default.svc.cluster.local
```

File: cluster/juju/layers/kubernetes-master/lib/charms/kubernetes/flagmanager.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
+from charmhelpers.core import unitdata
+
+
+class FlagManager:
+    '''
+    ServerOptsManager - A Python class for managing the flags to pass to an
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/lib/charms/kubernetes/flagmanager.py]: This was renamed?

File: cluster/juju/layers/kubernetes-master/lib/charms/kubernetes/flagmanager.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+            self.data = {}
+        else:
+            self.data = self.db.get(daemon)
+
+    def __save(self):
+        self.db.set(self.daemon, self.data)
+
+    def add(self, key, value, strict=False):
+        '''
+        Adds data to the map of values for the DockerOpts file.
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/lib/charms/kubernetes/flagmanager.py]: need to remove reference to DockerOpts file here.

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 5]
```diff
@@ -27,6 +35,12 @@ def install():
         hookenv.status_set('blocked', 'Missing kubernetes resource')
         return
 
+    # Handle null resource publication, we check if its filesize < 1mb
+    filesize = os.stat(archive).st_size
+    if filesize < 1000000:
+        hookenv.status_set('blocked', 'Missing kubernetes resource')
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: I would make this message different so we know the file was not complete. Suggest:
```
The kubernetes resource is not complete.
```

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 6]
```diff
@@ -98,11 +131,85 @@ def start_master(etcd):
...
+    ''' State guard to starting DNS '''
+    context = prepare_sdn_context(sdn_plugin)
+    render('kubedns-rc.yaml', '/etc/kubernetes/addons/kubedns-rc.yaml',
+           context)
+    render('kubedns-svc.yaml', '/etc/kubernetes/addons/kubedns-svc.yaml',
+           context)
+    # This should be auto-loaded by the addon manager, but it doesnt appear
+    # to do so.
+    launch_dns()
+    set_state('kube-dns.available')
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: This 'kube-dns.available' does not match the when_not start condition.

Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: also the launch_dns() method uses "kubedns.available" not "kube-dns.available" we should fix them all.

File: cluster/juju/layers/kubernetes-master/templates/kubedns-rc.yaml [hunk 7]
```diff
@@ -14,43 +14,48 @@
...
+        version: v19
         kubernetes.io/cluster-service: "true"
+      annotations:
+        scheduler.alpha.kubernetes.io/critical-pod: ''
+        scheduler.alpha.kubernetes.io/tolerations: '[{"key":"CriticalAddonsOnly", "operator":"Exists"}]'
     spec:
       containers:
       - name: kubedns
-        image: gcr.io/google_containers/kubedns-{{ arch }}:1.7
+        image: gcr.io/google_containers/kubedns-amd64:1.7
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/templates/kubedns-rc.yaml]: need to add the { arch } back in here

File: cluster/juju/layers/kubernetes-master/templates/kubedns-rc.yaml [hunk 8]
```diff
@@ -83,11 +87,12 @@ spec:
           name: dns-tcp-local
           protocol: TCP
       - name: dnsmasq
-        image: gcr.io/google_containers/kube-dnsmasq-{{ arch }}:1.3
+        image: gcr.io/google_containers/kube-dnsmasq-amd64:1.3
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/templates/kubedns-rc.yaml]: add the { arch } variable.

File: cluster/juju/layers/kubernetes-master/templates/kubedns-rc.yaml [hunk 9]
```diff
@@ -96,15 +101,19 @@ spec:
           name: dns-tcp
           protocol: TCP
       - name: healthz
-        image: gcr.io/google_containers/exechealthz-{{ arch }}:1.0
+        image: gcr.io/google_containers/exechealthz-amd64:1.1
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/templates/kubedns-rc.yaml]: use  { arch } here

File: cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py [hunk 10]
```diff
@@ -40,6 +49,12 @@ def install_kubernetes_components():
         hookenv.status_set('blocked', 'Missing kubernetes resource')
         return
 
+    # Handle null resource publication, we check if its filesize < 1mb
+    filesize = os.stat(kube_package).st_size
+    if filesize < 1000000:
+        hookenv.status_set('blocked', 'Missing kubernetes resource')
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: Would recommend using a different message for the incomplete resource. Suggest
```
The kubernetes resource is not complete.
```

**Comment by chuckbutler** [comment]
> PRs into smaller chucks

When i perfect the art of cloning, i will totally do this

**Pull request merged**